src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs

```diff
@@ -443,35 +447,40 @@ private static void CheckSnapshot(ITextSnapshot snapshot)
             // If we don't have any new tags, and there was nothing to invalidate, then we can 
             // keep whatever old tags we have without doing any additional work.
             if (noNewTags && noSpansToInvalidate)
-            {
                 return oldTagTree;
-            }
 
-            // We either have some new tags, or we have some tags to invalidate.
-            // First, determine which of the old tags we want to keep around.
-            var snapshot = noNewTags ? spansToInvalidate.First().Snapshot : newTags.First().Span.Snapshot;
-            var oldTagsToKeep = noSpansToInvalidate
-                ? oldTagTree.GetSpans(snapshot)
-                : GetNonIntersectingTagSpans(spansToInvalidate, oldTagTree);
-
-            // Then union those with the new tags to produce the final tag tree.
-            var finalTags = oldTagsToKeep.Concat(newTags);
-            return new TagSpanIntervalTree<TTag>(textBuffer, _dataSource.SpanTrackingMode, finalTags);
+            if (noSpansToInvalidate)
+            {
+                // If we have no spans to invalidate, then we can just keep the old tags and add the new tags.
+                var oldTagsToKeep = oldTagTree.GetSpans(newTags.First().Span.Snapshot);
```

Comment:
>.First() [](http://example.com/codeflow?start=63&length=8)

nit: [0]

Reply:
ArrayBuilder.First() is fine.  it's not linq :)